libs/langchain-community/src/vectorstores/upstash.ts

```diff
@@ -0,0 +1,241 @@
+ */
+export interface UpstashVectorLibArgs extends AsyncCallerParams {
+  index: UpstashIndex;
+}
+
+// eslint-disable-next-line @typescript-eslint/ban-types, @typescript-eslint/no-explicit-any
+export type UpstashMetadata = Record<string, any>;
+
+// eslint-disable-next-line @typescript-eslint/ban-types, @typescript-eslint/no-explicit-any
+export type UpstashQueryMetadata = UpstashMetadata & { documentContentLC: any };
+
+/**
+ * Type that defines the parameters for the delete method.
+ * It can either contain the target id(s) or the deleteAll config to reset all the vectors.
+ */
+export type UpstashDeleteParams =
+  | {
+      ids: string | string[];
+      deleteAll?: never;
+    }
+  | { deleteAll: boolean; ids?: never };
+
+const CONCURRENT_UPSERT_LIMIT = 1000;
+/**
+ * The main class that extends the 'VectorStore' class. It provides
+ * methods for interacting with Upstash index, such as adding documents,
+ * deleting documents, performing similarity search and more.
+ */
+export class UpstashVectorStore extends VectorStore {
+  index: UpstashIndex;
+
+  caller: AsyncCaller;
+
+  _vectorstoreType(): string {
+    return "upstash";
+  }
+
+  constructor(embeddings: EmbeddingsInterface, args: UpstashVectorLibArgs) {
+    super(embeddings, args);
+
+    this.embeddings = embeddings;
+
+    const { index, ...asyncCallerArgs } = args;
+
+    this.index = index;
+    this.caller = new AsyncCaller(asyncCallerArgs);
+  }
+
+  /**
+   * This method adds documents to Upstash database. Documents are first converted to vectors
+   * using the provided embeddings instance, and then upserted to the database.
+   * @param documents Array of Document objects to be added to the database.
+   * @param options Optional object containing array of ids for the documents.
+   * @returns Promise that resolves with the ids of the provided documents when the upsert operation is done.
+   */
+  async addDocuments(documents: Document[], options?: { ids?: string[] }) {
+    const texts = documents.map(({ pageContent }) => pageContent);
+
+    const embeddings = await this.embeddings.embedDocuments(texts);
+
+    return this.addVectors(embeddings, documents, options);
+  }
+
+  /**
+   * This method adds the provided vectors to Upstash database.
+   * @param vectors  Array of vectors to be added to the Upstash database.
+   * @param documents Array of Document objects, each associated with a vector.
+   * @param options Optional object containing the array of ids foor the vectors.
+   * @returns Promise that resolves with the ids of the provided documents when the upsert operation is done.
+   */
+  async addVectors(
+    vectors: number[][],
+    documents: Document[],
+    options?: { ids?: string[] }
+  ) {
+    const documentIds =
+      options?.ids ?? Array.from({ length: vectors.length }, () => uuid.v4());
+
+    const upstashVectors = vectors.map((vector, index) => {
+      const metadata = {
+        pageContentLC: documents[index].pageContent,
```

Comment:
Let's prefix this with an underscore to make clear it's only intended for internal use
```suggestion
        _pageContentLC: documents[index].pageContent,
```
Also, update ref's below